### PR TITLE
background mask option for pseudo labeler - unsupervised training

### PR DESCRIPTION
**Pull request by stmartineau99**
- added new class `NewPseudoLabeler` to subtract background from the pseudo labels if a background mask is given
- updated `get_unsupervised_loader` to handle 4 cases:

1. both sample mask and background mask are given, this will be be used by the train loader
2. sample mask only, this will be used by the validation loader
3. background mask only, not implemented error
4. neither mask given, use the default behavior

- created new classes to ensure that `get_unsupervised_loader` handles each case correctly: `DropChannel`, `ChannelWiseRawTransform`, `ChannelWiseAugmentations`
- updated `mean_teacher_adaptation` to use the `NewPseudoLabeler` if the background mask is given
- added subclass `NewMeanTeacherTrainer`, which drops the background mask from the teacher input after computing pseudo labels. it also drops the background mask from the student input since this is not used. This behavior only occurs when it recieves training data, and not for validation. 
- fixed the `NewPseudoLabeler` so that it behaves correctly when it recieves training and validation data, since in the case of validation there is no background mask channel.

### CODE REVIEW

File: synapse_net/training/domain_adaptation.py [hunk 1]
```diff
@@ -116,16 +199,22 @@ def mean_teacher_adaptation(
...
     loss_and_metric = self_training.DefaultSelfTrainingLossAndMetric()
-    
+
     unsupervised_train_loader = get_unsupervised_loader(
         data_paths=unsupervised_train_paths, 
         raw_key=raw_key, 
         patch_shape=patch_shape, 
         batch_size=batch_size, 
         n_samples=n_samples_train, 
+        sample_mask_paths=train_sample_mask_paths,
```
[the remark this answers is not in the document]
Comment (stmartineau99) [reply synapse_net/training/domain_adaptation.py:216]: some things got messed up when i was making this commit so it looks messy I think

File: synapse_net/training/domain_adaptation.py [hunk 2]
```diff
@@ -18,6 +20,184 @@
 from ..inference.inference import get_model_path, compute_scale_from_voxel_size
 from ..inference.util import _Scaler
 
+class NewPseudoLabeler(self_training.DefaultPseudoLabeler):
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:23]: I suggest we give this a more descriptive name: `PseudoLabelerWithBackgroundMask`

File: synapse_net/training/domain_adaptation.py [hunk 3]
```diff
@@ -18,6 +20,184 @@
...
+            if self.confidence_mask_channel is not None:
+                size = (pseudo_labels.shape[0], pseudo_labels.shape[1], *([-1] * (pseudo_labels.ndim - 2)))
+                label_mask = label_mask.expand(*size)
+        
+        if has_background_mask:
+            pseudo_labels = self._subtract_background(pseudo_labels, background_mask)
+
+        return pseudo_labels, label_mask
+
+class NewMeanTeacherTrainer(self_training.MeanTeacherTrainer):
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:100]: `MeanTeacherTrainerWithBackgroundMask`

File: synapse_net/training/domain_adaptation.py [hunk 4]
```diff
@@ -18,6 +20,184 @@
...
+    """Subclass of MeanTeacherTrainer, updated to handle cases where the background mask is provided. 
+    Once the pseudo labels are computed, the second channel of the teacher input is dropped, if it exists.
+    The second channel of the student input is also dropped, if it exists, since it is not needed for training.
+
+    Args:
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:105]: The Args in the doc string should match the actual arguments. To avoid too much duplication you could just use `**kwargs` for all the original arguments.

File: synapse_net/training/domain_adaptation.py [hunk 5]
```diff
@@ -18,6 +20,184 @@
...
+        unsupervised_val_loader: Optional[torch.utils.data.DataLoader] = None,
+        supervised_val_loader: Optional[torch.utils.data.DataLoader] = None,
+        supervised_loss: Optional[Callable] = None,
+        unsupervised_loss_and_metric: Optional[Callable] = None,
+        supervised_loss_and_metric: Optional[Callable] = None,
+        logger=SelfTrainingTensorboardLogger,
+        momentum: float = 0.999,
+        reinit_teacher: Optional[bool] = None,
+        sampler: Optional[Callable] = None,
+        **kwargs,
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:134]: I would suggest to change this to be
```python
**kwargs,
```

File: synapse_net/training/domain_adaptation.py [hunk 6]
```diff
@@ -18,6 +20,184 @@
...
+    ):
+        super().__init__(model, unsupervised_train_loader, unsupervised_loss, pseudo_labeler,
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:136]: and this just becomes 
```python
super().__init__(**kwargs)
```

File: synapse_net/training/domain_adaptation.py [hunk 7]
```diff
@@ -18,6 +20,184 @@
...
+    def _train_epoch_unsupervised(self, progress, forward_context, backprop):
+        self.model.train()
+
+        n_iter = 0
+        t_per_iter = time.time()
+
+        # Sample from both the supervised and unsupervised loader.
+        for xu1, xu2 in self.unsupervised_train_loader:
+            
+            # Assuming shape (B, C, D, H, W), only keep the first channel for xu2 (student input).
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:150]: I would add an assertion here that we have two channels. Otherwise the normal `MeanTeacherTrainer` should have been used and this may indicate some error in the code logic. So it's better to fail in that case.

File: synapse_net/training/domain_adaptation.py [hunk 8]
```diff
@@ -36,13 +216,14 @@ def mean_teacher_adaptation(
...
-    val_mask_paths: Optional[Tuple[str]] = None,
+    train_sample_mask_paths: Optional[Tuple[str]] = None,
+    val_sample_mask_paths: Optional[Tuple[str]] = None,
+    train_background_mask_paths: Optional[Tuple[str]] = None,
     patch_sampler: Optional[callable] = None,
     pseudo_label_sampler: Optional[callable] = None,
     device: int = 0,
 ) -> None:
-    """Run domain adaptation to transfer a network trained on a source domain for a supervised
+    """Run domain adapation to transfer a network trained on a source domain for a supervised
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:226]:
```suggestion
    """Run domain adaptation to transfer a network trained on a source domain for a supervised
```

File: synapse_net/training/domain_adaptation.py [hunk 9]
```diff
@@ -109,24 +294,29 @@ def mean_teacher_adaptation(
         if os.path.isdir(source_checkpoint):
             model = torch_em.util.load_model(source_checkpoint)
         else:
-            model = torch.load(source_checkpoint, weights_only=False)
+            model = torch.load(source_checkpoint)
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:297]:
```suggestion
            model = torch.load(source_checkpoint, weights_only=False)
```

Comment (constantinpape) [reply synapse_net/training/domain_adaptation.py:297]: We need to keep `weights_only=False` here to be compatible with recent PyTorch versions. If you removed this to fix an error in PyTorch then please update your PyTorch.

File: synapse_net/training/domain_adaptation.py [hunk 10]
```diff
@@ -109,24 +294,29 @@ def mean_teacher_adaptation(
...
         reinit_teacher = False
 
     optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)
     scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode="min", factor=0.5, patience=5)
 
     # self training functionality
-    pseudo_labeler = self_training.DefaultPseudoLabeler(confidence_threshold=confidence_threshold)
+    if train_background_mask_paths is not None:
+        pseudo_labeler = NewPseudoLabeler(confidence_threshold=confidence_threshold, background_mask_channel=1)
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:305]: Rename to `PseudoLabelerWithBackgroundMask`, see comment above.

File: synapse_net/training/domain_adaptation.py [hunk 11]
```diff
@@ -109,24 +294,29 @@ def mean_teacher_adaptation(
...
-            model = torch.load(source_checkpoint, weights_only=False)
+            model = torch.load(source_checkpoint)
         reinit_teacher = False
 
     optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)
     scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode="min", factor=0.5, patience=5)
 
     # self training functionality
-    pseudo_labeler = self_training.DefaultPseudoLabeler(confidence_threshold=confidence_threshold)
+    if train_background_mask_paths is not None:
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:304]: You now also have to select the correct trainer class here. In the first case (i.e. `train_background_mask_paths is not None`), you have to define `trainer_class = MeanTeacherTrainerWithBackgroundMask`

File: synapse_net/training/domain_adaptation.py [hunk 12]
```diff
@@ -109,24 +294,29 @@ def mean_teacher_adaptation(
...
+        pseudo_labeler = NewPseudoLabeler(confidence_threshold=confidence_threshold, background_mask_channel=1)
+    else:
```
Comment (constantinpape) [inline synapse_net/training/domain_adaptation.py:306]: And in this case set `trainer_class = self_training.MeanTeacherTrainer`

Comment (constantinpape) [reply synapse_net/training/domain_adaptation.py:306]: And in line 348 you have to use `trainer_class`.

**Review by constantinpape — COMMENTED** [review]
This looks very good now!  The only things to address:
- You are not actually using the new class for the mean teacher trainer.
- Some cosmetics.

See details in the comments.

**Comment by stmartineau99** [comment]
I think some changes for these files did not get merged to the main branch: 
- training/semisupervised_training.py
- training/domain_adaptation.py

**Comment by stmartineau99** [comment]
I added gpu device arguments to some other files, and my personal training scripts.

**Review by stmartineau99 — COMMENTED** [review]
- mainly changes to semisupervised_training.py - get_unsupervised_loader()
- changes to domain_adaptation.py - responded to your last comment about asserting dims.

**Comment by constantinpape** [comment]
Is this still relevant @stmartineau99 ? Or is this now addressed by #165 ?

**Comment by stmartineau99** [comment]
> Is this still relevant @stmartineau99 ? Or is this now addressed by #165 ?

I closed this PR, its not needed.
